src/fd.cpp

```diff
@@ -0,0 +1,248 @@
+  FD_ZERO(&readfds);
+  FD_ZERO(&writefds);
+  FD_ZERO(&exceptfds);
+  const int max_fd = *std::max_element(args->fds->begin(), args->fds->end());
+  struct timeval tv;
+
+  do {
+    for (int i = 0; i < args->rfds; i++) {
+      FD_SET(fds[i], &readfds);
+    }
+    for (int i = args->rfds; i < (args->rfds + args->wfds); i++) {
+      FD_SET(fds[i], &writefds);
+    }
+    for (int i = args->rfds + args->wfds; i < args->num_fds; i++) {
+      FD_SET(fds[i], &exceptfds);
+    }
+    tv.tv_sec = (repeats ? LATER_INTERVAL : timeoutms) / 1000;
+    tv.tv_usec = ((repeats ? LATER_INTERVAL : timeoutms) % 1000) * 1000;
+
+    result = select(max_fd + 1, &readfds, &writefds, &exceptfds, &tv);
+
+    if (args->flag->load()) goto callback;
+    if (result) break;
+  } while (infinite || (repeats-- && (timeoutms -= LATER_INTERVAL)));
+
+  if (result > 0) {
+    for (int i = 0; i < args->rfds; i++) {
+      fds[i] = FD_ISSET(fds[i], &readfds);
+    }
+    for (int i = args->rfds; i < (args->rfds + args->wfds); i++) {
+      fds[i] = FD_ISSET(fds[i], &writefds);
+    }
+    for (int i = args->rfds + args->wfds; i < args->num_fds; i++) {
+      fds[i] = FD_ISSET(fds[i], &exceptfds);
+    }
+  } else if (result == 0) {
+    std::memset(fds, 0, args->num_fds * sizeof(int));
+  } else {
+    for (int i = 0; i < args->num_fds; i++) {
+      fds[i] = R_NaInt;
+    }
+  }
+
+#else
+
+  std::vector<struct pollfd> pollfds;
+  pollfds.reserve(args->num_fds);
+  struct pollfd pfd;
+  for (int i = 0; i < args->rfds; i++) {
+    pfd.fd = fds[i];
+    pfd.events = POLLIN;
+    pfd.revents = 0;
+    pollfds.push_back(pfd);
+  }
+  for (int i = args->rfds; i < (args->rfds + args->wfds); i++) {
+    pfd.fd = fds[i];
+    pfd.events = POLLOUT;
+    pfd.revents = 0;
+    pollfds.push_back(pfd);
+  }
+  for (int i = args->rfds + args->wfds; i < args->num_fds; i++) {
+    pfd.fd = fds[i];
+    pfd.events = 0;
+    pfd.revents = 0;
+    pollfds.push_back(pfd);
+  }
+
+  do {
+    result = POLL_FUNC(pollfds.data(), args->num_fds, repeats ? LATER_INTERVAL : timeoutms);
+    if (args->flag->load()) goto callback;
+    if (result) break;
+  } while (infinite || (repeats-- && (timeoutms -= LATER_INTERVAL)));
+
+  if (result > 0) {
+    for (int i = 0; i < args->rfds; i++) {
+      fds[i] = pollfds[i].revents == 0 ? 0 : pollfds[i].revents & POLLIN ? 1: R_NaInt;
+    }
+    for (int i = args->rfds; i < (args->rfds + args->wfds); i++) {
+      fds[i] = pollfds[i].revents == 0 ? 0 : pollfds[i].revents & POLLOUT ? 1 : R_NaInt;
+    }
+    for (int i = args->rfds + args->wfds; i < args->num_fds; i++) {
+      fds[i] = pollfds[i].revents != 0;
+    }
+  } else if (result == 0) {
+    std::memset(fds, 0, args->num_fds * sizeof(int));
+  } else {
+    for (int i = 0; i < args->num_fds; i++) {
+      fds[i] = R_NaInt;
+    }
+  }
+
+#endif // POLLIN
+
+  callback:
+  callbackRegistryTable.scheduleCallback(later_callback, static_cast<void *>(argsptr.release()), 0, args->loop);
```

Comment:
Instead of exposing `callbackRegistryTable`, could you use `execLaterNative2` which is already exported?

Reply:
Amended in [989e33e](https://github.com/r-lib/later/pull/190/commits/989e33ec4bae6b9568ee26dc6810c75d957e34ee).

I actually used `execLaterNative2()` to start with, but it also calls `ensureInitialized()`, which I accept doesn't really make a difference here, but conceptually it seemed like something for external users rather than a function from within the package.

Reply:
Oh, that's a good point. I think I still prefer it like this for now. (I could see all the code moving around a bit if we want to introduce a native `later_fd` someday.)

Reply:
You raise a really good point about the native version of the function.

I'm not one to leave things half-done, so I decided to give myself 2 hours this morning to try and attempt this, and I think I've just about succeeded! Will open up a new PR.